scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +741,336 @@ class ProcessMonitor(ScalyrMonitor):
+            # for all the absolute metrics, decrease the count that the dead processes accounted for
+            del self.__metrics_history[_pid_to_remove]
+
+        # if no processes are running, there is no reason to report the running metric data
+        if not running_pids:
+            self.__aggregated_metrics = {}
+
+    def get_pids_from_ps(self, subprocesses=False, ppid=None):
+        """
+        Get the list of pids running, or get the list of child processes given a parent process id
+        :param subprocesses: Get subprocesses?
+        :param ppid: parent process id, only if subprocesses is True
+        :type subprocesses: bool
+        :type ppid: int
+        :return: list of process ids
+        """
 
-        for gather in self.__gathers:
-            gather.run_single_cycle()
+        cmd = ['ps', 'ax', '-o', 'pid,command']
+        if subprocesses:
+            cmd = ["ps", "--ppid", str(ppid), "-o", "pid,command"]
+        sub_proc = Popen(cmd, shell=False, stdout=PIPE)
+        lines = sub_proc.stdout.readlines()
+        all_pids = []
+        try:
+            for line in lines:
+                line = line.strip()
+                if line.find(' ') > 0:
+                    pid = line[:line.find(' ')]
+                    if pid.lower() == 'pid':
+                        continue
+                    pid = int(pid)
+                    line = line[(line.find(' ') + 1):]
+                    if not subprocesses:
+                        if re.search(self.__commandline_matcher, line) is not None:
+                            all_pids.append(pid)
+                    else:
+                        all_pids.append(pid)
+        finally:
+            if sub_proc is not None:
+                sub_proc.wait()
+        return all_pids
+
+    def get_active_matched_pids(self, check_poll=True):
+        """
+        Get the active pids that match the commandline match expression.
+        If there is no need to poll, then return the currently monitored pids
+        @:param check_poll: Should check for new matched processes?
+        @:param include_child_processes: Should include child processes of matching processes?
+        @:type check_poll: bool
+        @:type include_child_processes: bool
+        :return: List of running process IDs
+        """
 
-    def __select_process(self):
-        """Returns the proces id of a running process that fulfills the match criteria.
+        if check_poll:
+            if not self.should_poll_for_processes():
+                return self.__pids
 
-        This will either use the commandline matcher or the target pid to find the process.
-        If no process is matched, None is returned.
+        matching_pids = self.get_pids_from_ps()
+        if self.__include_child_processes:
+            for matching_pid in matching_pids:
+                matching_pids.extend(
+                    self.get_pids_from_ps(subprocesses=True, ppid=matching_pid)
```

Comment:
The drawback here is you keep having to invoke `ps aux` for every child process.